R/assert.R

```diff
@@ -95,8 +95,14 @@ vec_is <- function(x, ptype = NULL, size = NULL) {
 
 is_same_type <- function(x, ptype) {
   if (is_partial(ptype)) {
-    ptype <- vec_type_common(x, ptype)
+    ptype <- maybe(vec_type_common(x, ptype), "vctrs_error_incompatible_type")
```

Comment:
Given that we're not fully committed to this pattern, I think it would be better to inline the `tryCatch()` here and avoid any abstraction to hide the ugliness of the approach.